src/blog/fearless-structured-concurrency.vmd

```diff
@@ -0,0 +1,542 @@
+
+
+```rust
+let m = Mutex::new(100i64);
+
+let escaped_value =
+  {
+    let counter = &mut *m.lock().unwrap();
+    *counter = *counter + 1i64;
+    let new_value = *counter + 100
+    new_value
+  };
+
+println!("{:?}", escaped_value);
+```
+
+
+Rust uses the type system to track where a value came from. [# This part of the type system is called the "borrow checker". It's actually much more complicated and powerful than alluded to here.]
+
+
+<slice/>
+
+
+# How do we combine them?
+
+We can make structured concurrency fearless by:
+
+ 1. Start with structured concurrency.
+ 1. We *can't* modify values created *outside* a parallel block.
+ 1. ...except for mutexes, we can always lock them.
+
+
+Our C program _almost_ follows these rules, except it violates rule 2; we modified the `results` array inside the parallel block:
+
+
+```c
+int exponent = 3;
+
+int results[5];
+
+// Launch some threads and run in parallel!
+#pragma omp parallel for
+for (int i = 0; i < 5; i++) {
+  results[i] = pow(i, exponent);
+}
+```
+
+
+
+But Vale has a `foreach` loop that can accumulate each iteration's result, and produce an array: [# For those unfamiliar, Vale is a language that ... (put stuff here)]
+
+
+```
+exponent = 3;
+
+results =
+  parallel foreach 0..5 as i { «executor»
+    = pow(i, exponent); «blockresult1»
+  }
+```
+
+
+
+There! Now the loop doesn't modify anything created outside the block.
+
+
+And just because we can, let's modify something created _inside_ the block:
+
+
+```
+fn main() export {
+  exponent = 3;
+
+  results =
+    parallel foreach 0..5 as i {
+      a = pow(i, exponent);
+      set a = a + 4; // We can modify!
+      = a; «blockresult2»
+    }
+}
+```
+
+You can see how the compiler can enforce that we only modify things that came from inside the block.
+
+
+The type system also needs to track what we can modify and what we can't. For that, we use `'r`.
+
+
+Note how in this snippet, `blur`'s first parameter's type is `'r Array<Array<int>>`. The `'r` enforces that the function won't modify the array from this function. [# This is similar to Rust's lifetimes, more on that below.]
+
+
+```
+fn main() export {
+  // Makes a 2D array of ints
+  input = Array(10, x => Array(10, y => x * 10 + y));
+
+  results =
+    parallel foreach 0..10 as x {
+      = map 0..10 as y {
+          = blur(input, x, y);
+        }
+    }
+}
+
+fn blur(input 'r &Array<Array<int>>, x int, y int) int {
+  ... // Loop over the 3x3 ints around x,y in input and return the average
+}
+```
+
+
+
+That `'r` means we're pointing into an *immutable region*. More on that below!
+
+
+<slice>
+#executor: The `parallel` block will by default spawn `num_cores * 2` threads, but it can also be specified via `parallel(num_threads)` or `parallel(an_executor)`.
```

Comment:
"The `parallel` block uses the default executor, but it can also be explicitly overloaded with `parallel(num_threads)` or `parallel(an_executor).`"

Probably best to keep the details vague, and focus on the semantics.

Reply:
Done